### PR TITLE
wireless/bcm43xxx: sort scan result by rssi

### PR DESCRIPTION
**Pull request by anchao**
## Summary

wireless/bcm43xxx: sort scan result by rssi
  
1. Replace SCAN_RESULT_SIZE to SCAN_RESULT_ENTRIES
2. filter scan result with better rssi
3. Sort scan result by rssi

## Impact

N/A

## Testing

bcm43013 scan

### CODE REVIEW

**Review by acassis — APPROVED** [review]

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c [hunk 1]
```diff
@@ -666,300 +669,235 @@ void bcmf_wl_scan_event_handler(FAR struct bcmf_dev_s *priv,
...
-
-      /* Special processing for iw_point, set offset in pointer field */
-
-      iwe->u.essid.pointer = (FAR void *)sizeof(iwe->u.essid);
-      memcpy(&iwe->u.essid + 1, bss->SSID, essid_len);
-
-      priv->scan_result_size += BCMF_IW_EVENT_SIZE(essid)+essid_len_aligned;
-      result_size -= BCMF_IW_EVENT_SIZE(essid)+essid_len_aligned;
+      curr = &priv->scan_result[priv->scan_result_entries];
+      memcpy(curr, bss, sizeof(wl_bss_info_t));
```
Comment (pkarashchenko) [inline drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c:737]:
```suggestion
      memcpy(curr, bss, sizeof(*curr));
```
Just to follow the same style as at line 728 and 712

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c [hunk 2]
```diff
@@ -56,7 +56,7 @@
 #define DOT11_BSSTYPE_ANY      2
 #define BCMF_SCAN_TIMEOUT_TICK (5*CLOCKS_PER_SEC)
 #define BCMF_AUTH_TIMEOUT_MS   20000  /* was 10000 */
```
Comment (pkarashchenko) [inline drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c:58]:
```suggestion
#define DOT11_BSSTYPE_ANY         2
#define BCMF_SCAN_TIMEOUT_TICK    (5*CLOCKS_PER_SEC)
#define BCMF_AUTH_TIMEOUT_MS      20000  /* was 10000 */
```

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c (hunk 1 above)
Comment (anchao) [reply drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c:737]: done

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c (hunk 2 above)
Comment (anchao) [reply drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c:58]: done

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c [hunk 3]
```diff
@@ -666,300 +669,247 @@ void bcmf_wl_scan_event_handler(FAR struct bcmf_dev_s *priv,
...
+  bcmf_hexdump((FAR uint8_t *)event, event_len, (unsigned long)event);
+}
 
-      priv->scan_result_size += BCMF_IW_EVENT_SIZE(data);
-      result_size -= BCMF_IW_EVENT_SIZE(data);
+static FAR char *
+bcmf_wl_iwe_add_event(FAR char *stream, FAR char *stop,
+                      FAR struct iw_event *iwe, int event_len)
+{
+  if (stream + event_len > stop)
```
Comment (xiaoxiang781216) [inline drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c:790]: why need check again? Line 833 already handle this case.

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c [hunk 4]
```diff
@@ -666,300 +669,247 @@ void bcmf_wl_scan_event_handler(FAR struct bcmf_dev_s *priv,
...
-    }
+      /* Copy BSSID */
 
-wl_escan_result_processed:
+      iwe = (FAR struct iw_event *)start;
+      iwe->cmd = SIOCGIWAP;
+      iwe->u.ap_addr.sa_family = ARPHRD_ETHER;
+      memcpy(&iwe->u.ap_addr.sa_data,
+             info->BSSID.ether_addr_octet, IFHWADDRLEN);
+      start = bcmf_wl_iwe_add_event(start, stop, iwe, IW_EV_LEN(ap_addr));
```
Comment (xiaoxiang781216) [inline drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c:867]: let's use iwe directly and remove start/stop

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c [hunk 5]
```diff
@@ -623,16 +623,19 @@ void bcmf_wl_auth_event_handler(FAR struct bcmf_dev_s *priv,
...
-  struct wl_bss_info *bss;
-  unsigned int bss_info_len;
+  FAR struct wl_escan_result *result;
   unsigned int escan_result_len;
-  unsigned int bss_count = 0;
+  FAR struct wl_bss_info *curr;
+  FAR struct wl_bss_info *bss;
+  uint32_t event_len;
+  int16_t worst_rssi;
+  int worst_entries;
```
[the remark this answers is not in the document]
Comment (anchao) [reply drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c:635]: done

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c [hunk 6]
```diff
@@ -666,300 +669,247 @@ void bcmf_wl_scan_event_handler(FAR struct bcmf_dev_s *priv,
...
+          /* Check if current bss AP is not already detected */
 
-          if (iwe->cmd == SIOCGIWAP)
+          if (memcmp(&curr->BSSID, &bss[i].BSSID,
+                     sizeof(curr->BSSID)) == 0)
             {
-              if (memcmp(&iwe->u.ap_addr.sa_data,
-                         bss->BSSID.ether_addr_octet,
-                         sizeof(bss->BSSID.ether_addr_octet)) == 0)
+              /* Replace the duplicate entries if rssi is
```
[the remark this answers is not in the document]
Comment (anchao) [reply drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c:718]: done

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c (hunk 4 above)
Comment (anchao) [reply drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c]: Done

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c (hunk 3 above)
Comment (anchao) [reply drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c]: Done

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c [hunk 7]
```diff
@@ -666,300 +669,247 @@ void bcmf_wl_scan_event_handler(FAR struct bcmf_dev_s *priv,
...
+            }
 
-      while (priv->scan_result_size - check_offset
-                                     >= offsetof(struct iw_event, u))
-        {
-          iwe = (struct iw_event *)&priv->scan_result[check_offset];
+          /* Check if current bss AP is not already detected */
 
-          if (iwe->cmd == SIOCGIWAP)
+          if (memcmp(&curr->BSSID, &bss[i].BSSID,
```
[the remark this answers is not in the document]
Comment (anchao) [reply drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c:707]: Done

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c [hunk 8]
```diff
@@ -666,300 +669,247 @@ void bcmf_wl_scan_event_handler(FAR struct bcmf_dev_s *priv,
...
-          goto exit_invalid_frame;
-        }
-
-      /* Append current bss_info to priv->scan_results
-       * FIXME protect this against race conditions
-       */
+          curr = &priv->scan_result[j];
 
-      /* Check if current bss AP is not already detected */
+          /* Find worst rssi and mark the entries */
```
[the remark this answers is not in the document]
Comment (anchao) [reply drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c]: Done

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c [hunk 9]
```diff
@@ -666,300 +669,247 @@ void bcmf_wl_scan_event_handler(FAR struct bcmf_dev_s *priv,
...
       result->buflen < sizeof(struct wl_escan_result))
     {
       goto exit_invalid_frame;
     }
 
-  /* wl_escan_result structure contains a wl_bss_info field */
-
-  len = result->buflen - sizeof(struct wl_escan_result)
-                       + sizeof(struct wl_bss_info);
+  bss = result->bss_info;
```
[the remark this answers is not in the document]
Comment (anchao) [reply drivers/wireless/ieee80211/bcm43xxx/bcmf_driver.c]: done

**Review by xiaoxiang781216 — APPROVED** [review]
LGTM.

**Pull request merged**